File: src/main/java/me/zeroeightsix/kami/module/modules/experimental/FancyChat.java
package me.zeroeightsix.kami.module.modules.experimental;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.module.modules.chat.CustomChat;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.client.CPacketChatMessage;

import java.util.Random;

import static me.zeroeightsix.kami.util.InfoCalculator.isNumberEven;

/**
 * @author S-B99
 * Updated by S-B99 on 12/03/20
 */
@Module.Info(name = "FancyChat", category = Module.Category.EXPERIMENTAL, description = "Makes messages you send fancy", showOnArray = Module.ShowOnArray.OFF)
public class FancyChat extends Module {
    private Setting<Mode> modeSetting = register(Settings.e("Mode", Mode.MOCKING));
    private Setting<Boolean> randomSetting = register(Settings.booleanBuilder("Random Case").withValue(true).withVisibility(v -> modeSetting.getValue().equals(Mode.MOCKING)).build());
    private Setting<Boolean> commands = register(Settings.b("Commands", false));

    private enum Mode { UWU, LEET, MOCKING }
    private static Random random = new Random();

    private String getText(Mode t, String s) {
        switch (t) {
            case UWU:
                return "Error: UWU mode is still experimental";
            case LEET:
                return leetConverter(s);
            case MOCKING:
                return mockingConverter(s);
            default:
                return "";
        }
    }

    // rules:
    // ove > uv
    // l > w
    //
//    private String uwuifier(String input) {
//        input.replace("a", "b");
//    }

    private String leetConverter(String input) {
        StringBuilder message = new StringBuilder();
        for (int i = 0 ; i < input.length() ; i++) {
            String inputChar = input.charAt(i) + "";
            inputChar = inputChar.toLowerCase();
            inputChar = leetSwitch(inputChar);
            message.append(inputChar);
        }
        return message.toString();
    }

    private String mockingConverter(String input) {
        StringBuilder message = new StringBuilder();
        for (int i = 0 ; i < input.length() ; i++) {
            String inputChar = input.charAt(i) + "";

            int rand = 0;
            if (randomSetting.getValue()) rand = random.nextBoolean() ? 1 : 0;

            if (!isNumberEven(i + rand)) inputChar = inputChar.toUpperCase();
            else inputChar = inputChar.toLowerCase();
            message.append(inputChar);
        }
        return message.toString();
    }

    @EventHandler
    public Listener<PacketEvent.Send> listener = new Listener<>(event -> {
        if (event.getPacket() instanceof CPacketChatMessage) {
            String s = ((CPacketChatMessage) event.getPacket()).getMessage();
            if (!commands.getValue() && isCommand(s)) return;
            s = getText(modeSetting.getValue(), s);
            if (s.length() >= 256) s = s.substring(0, 256);
            ((CPacketChatMessage) event.getPacket()).message = s;
        }
    });

    private boolean isCommand(String s) {
        for (String value : CustomChat.cmdCheck) {
            if (s.startsWith(value)) return true;
        }
        return false;
    }

    private String leetSwitch(String i) {
        switch (i) {
            case "a":
                return "4";
            case "e":
                return "3";
            case "g":
                return "6";
            case "l":
            case "i":
                return "1";
            case "o":
                return "0";
            case "s":
                return "$";
            case "t":
                return "7";
            default: return i;
        }
    }
}

File: src/main/kotlin/org/kamiblue/client/module/modules/chat/FancyChat.kt
package org.kamiblue.client.module.modules.chat

import org.kamiblue.client.manager.managers.MessageManager.newMessageModifier
import org.kamiblue.client.module.Category
import org.kamiblue.client.module.Module
import org.kamiblue.client.util.text.MessageDetection
import org.kamiblue.commons.utils.MathUtils
import kotlin.math.min

internal object FancyChat : Module(
    name = "FancyChat",
    category = Category.CHAT,
    description = "Makes messages you send fancy",
    showOnArray = false,
    modulePriority = 100
) {
    private val uwu = setting("uwu", true)
    private val leet = setting("1337", false)
    private val mock = setting("mOcK", false)
    private val green = setting(">", false)
    private val blue = setting("`", false)
    private val randomSetting = setting("Random Case", true, { mock.value })
    private val commands = setting("Commands", false)
    private val spammer = setting("Spammer", false)

    private val modifier = newMessageModifier(
        filter = {
            (commands.value || MessageDetection.Command.ANY detectNot it.packet.message)
                && (spammer.value || it.source !is Spammer)
        },
        modifier = {
            val message = getText(it.packet.message)
            message.substring(0, min(256, message.length))
        }
    )

    init {
        onEnable {
            modifier.enable()
        }

        onDisable {
            modifier.disable()
        }
    }

    private fun getText(s: String): String {
        var string = s
        if (uwu.value) string = uwuConverter(string)
        if (leet.value) string = leetConverter(string)
        if (mock.value) string = mockingConverter(string)
        if (green.value) string = greenConverter(string)
        if (blue.value) string = blueConverter(string)
        return string
    }

    private fun greenConverter(input: String): String {
        return "> $input"
    }

    private fun blueConverter(input: String): String {
        return "`$input"
    }

    override fun getHudInfo(): String {
        val returned = StringBuilder()
        if (uwu.value) {
            returned.append("uwu")
        }
        if (leet.value) {
            returned.append(" 1337")
        }
        if (mock.value) {
            returned.append(" mOcK")
        }
        if (green.value) {
            returned.append(" >")
        }
        if (blue.value) {
            returned.append(" `")
        }
        return returned.toString()
    }

    private fun leetConverter(input: String): String {
        val message = StringBuilder()
        for (element in input) {
            var inputChar = element.toString() + ""
            inputChar = inputChar.toLowerCase()
            inputChar = leetSwitch(inputChar)
            message.append(inputChar)
        }
        return message.toString()
    }

    private fun mockingConverter(input: String): String {
        val message = StringBuilder()
        for (i in input.indices) {
            var inputChar = input[i].toString() + ""
            val rand = if (randomSetting.value) (0..1).random() else 0
            inputChar = if (!MathUtils.isNumberEven(i + rand)) inputChar.toUpperCase() else inputChar.toLowerCase()
            message.append(inputChar)
        }
        return message.toString()
    }

    private fun uwuConverter(input: String): String {
        var lInput = input
        lInput = lInput.replace("ove", "uv")
        lInput = lInput.replace("the", "da")
        lInput = lInput.replace("is", "ish")
        lInput = lInput.replace("r", "w")
        lInput = lInput.replace("ve", "v")
        lInput = lInput.replace("l", "w")
        return lInput
    }

    private fun leetSwitch(i: String): String {
        return when (i) {
            "a" -> "4"
            "e" -> "3"
            "g" -> "6"
            "l", "i" -> "1"
            "o" -> "0"
            "s" -> "$"
            "t" -> "7"
            else -> i
        }
    }
}
